### ISSUE TITLE
Implement QAT at TorchTitan

### ISSUE THREAD

**Issue by mori360** [pytorch/torchtitan#2742]
Post-training quantization (PTQ) is the standard approach for deploying quantized LLMs, but it degrades model quality because the model was never exposed to quantization noise  during training. QAT addresses this by inserting fake quantization (simulated quantize/dequantize round-trips) during training, so the model learns weights that are inherently robust to quantization. Combined with LoRA for parameter-efficient fine-tuning, QAT+LoRA enables quantization-friendly adaptation without full-parameter training costs.

Reference link:
https://pytorch.org/blog/quantization-aware-training/
https://pytorch.org/blog/quantization-aware-training-in-torchao-ii/

Current work:
We have the LoRA and QAT draft in https://github.com/pytorch/torchtitan/pull/2484, https://github.com/pytorch/torchtitan/pull/2488, results below are based on the implementation in the link.

Results: 
Loading pre-trained Llama 3.1 8B from huggingface, finetune 100 steps on 8GPU, C4, lm-eval benchmarks
Then have experiments as:
(blue): LoRA finetune, full-precision eval
(orange): LoRA finetune, PTQ eval
(green): QAT-LoRA finetune, post training quantization eval
1. There's accuracy drop with PTQ (from blue to orange)
2.  QAT benefit scales with quantization aggressiveness. For int8act+int4weight, QAT+LoRA recovers 29-67% of the PTQ accuracy loss across benchmarks.
3. QAT-LoRA adds ~40% overhead to LoRA, with torch.compile, QAT overhead drops from +41% to +12%.
<img width="2380" height="771" alt="Image" src="https://github.com/user-attachments/assets/6d93d889-63e5-4298-836f-9124a34fc88c" />

Then comes to the problem on QAT-LoRA design. As mention above that finetuning with noise from quantization could get better PTQ eval results, but we wanted to test it, as well as to test the coverage of other models.
Then we test on Qwen 0.8B model, lm-eval benchmarks
The results here shows:
1. Apply QAT to adapters has a big cost, as we has 2 more converters for each Linear
2. In the given model, there's no drop for non-QAT at adapter, actually there's slight improvements in some datasets(need more work here to double check).
<img width="2946" height="1857" alt="Image" src="https://github.com/user-attachments/assets/d5ebdaf2-b427-4ede-903c-008f0dfb5953" />

TODO:
1. We need to test more schemes available from torchao, have some initial guidance on selection, as well as mode tests
2.  During testing, we find there's space for hyper-parameter during QAT finetune, need more experiments to find the sweet point there.
3. More models testing, also to double check whether there's no need for QAT on converters.

**Comment by mori360** [comment]
We have more experiments on Qwen3 and Llama3 8b, showing the consistent results on QAT that performance between than plan PTQ.
Also we have some research on whether we should enable qat to lora adapter as well, seems we could support qat to both lora_a and lora_b, but could skip lora_b is rank < group_size that we cannot apply qat to lora_b.
<img width="2447" height="2098" alt="Image" src="https://github.com/user-attachments/assets/47e50564-73e8-4e26-abf8-2b09f419115f" />

<img width="2461" height="2098" alt="Image" src="https://github.com/user-attachments/assets/05899009-6436-4956-b908-9abfd0bc131a" />

Discussion:
A. Currently we focus on the order: Linear -> FakeQuantizedLinear -> LoRAFakeQuantizedLinear 

Cons: we need to apply qat twice, first time to the origin weight, second time to the adapters
Strength: 
1. QAT(Linear) would miss the other components excepting weight and bias, so that adapter info would be missed.
2. We apply QAT to lora_b specially with condition, so that if we apply qat second, that might raise error directly, which is unnecessary.

B. what's format should we support in QAT checkpoint.
Here are the uscases I could image:
1. we finetune with QAT and get the updated model state, and hope to eval the model state(or upload to hugging face)
2. during qat finetune, we want to continue finetuning, so that load the qat state and continue
Based on these 2 cases, we might currently need both the dcp and hf format, further support will be considered in the checkpoint design that we currenly realize there are more asks to checkpoint when we plan to  support LoRA and QAT, so plans will be shared in other issues next week.

C. hyperparam tuning
Actually we find that the default hyper-param could not achieve the best performance, even make it worse. For example for the float8 scheme, we find that learning rate at 1e-5 could achieve better convergence and eval results.
But it's huge work for hyperparam tuning, so we will remain it as future work until LoRA component is available and fixed.

D. coverage to more models
We currently have experiments on Llama3.1 and Qwen3. Other models containing moe, which is not supported well in the current lora draft, so we need to device the lora-moe plan, then cover more models.
